control-plane/pkg/reconciler/sink/kafka_sink.go

```diff
@@ -71,19 +80,34 @@ func (r *Reconciler) reconcileKind(ctx context.Context, ks *eventing.KafkaSink)
 	}
 	statusConditionManager.DataPlaneAvailable()
 
+	if ks.GetStatus().Annotations == nil {
+		ks.GetStatus().Annotations = make(map[string]string, 1)
+	}
+
 	if ks.Spec.NumPartitions != nil && ks.Spec.ReplicationFactor != nil {
-		topicConfig := &kafka.TopicConfig{
-			TopicDetail: sarama.TopicDetail{
-				NumPartitions:     *ks.Spec.NumPartitions,
-				ReplicationFactor: *ks.Spec.ReplicationFactor,
-			},
-			BootstrapServers: ks.Spec.BootstrapServers,
-		}
+
+		ks.GetStatus().Annotations[TopicOwnerAnnotation] = ControllerTopicOwner
+
+		topicConfig := topicConfigFromSinkSpec(&ks.Spec)
 
 		topic, err := r.ClusterAdmin.CreateTopic(logger, ks.Spec.Topic, topicConfig)
 		if err != nil {
 			return statusConditionManager.FailedToCreateTopic(topic, err)
 		}
+	} else {
+
+		// If the topic is externally managed, we need to make sure that the topic exists and it's valid.
+
+		ks.GetStatus().Annotations[TopicOwnerAnnotation] = ExternalTopicOwner
+
+		isPresentAndValid, err := r.ClusterAdmin.IsTopicPresentAndValid(ks.Spec.Topic, ks.Spec.BootstrapServers)
+		if err != nil {
+			return statusConditionManager.TopicNotPresentOrInvalidErr(err)
+		}
+		if !isPresentAndValid {
+			// The topic might be invalid.
+			return statusConditionManager.TopicNotPresentOrInvalid()
+		}
 	}
 	statusConditionManager.TopicCreated(ks.Spec.Topic)
```

Comment:
More than using always the same status, can we have a status for "topic created owned by controller" and "topic ready owned by external"? It provides more verbosity of what the controller does and maybe we can use it even to distinguish owned/not owned topics, avoiding the annotation

Reply:
The annotation isn't designed for us to distinguish the two cases since `numPartitions` and `replicationFactor` are immutable we can always infer who owns the topic.
The reason for using the annotation is to allow users or machines (since it's structured) using the status regardless of the internal implementation that might vary, and it's an information that is already in the status, so I'm not sure adding verbosity in the status is a pattern I've seen in k8s, it seems more appropriate for a log message. WDYT?

Reply:
On the contrary, to me this really sounds like a status thing: `TopicAlreadyPresent` vs `NewTopicCreated`. @matzew we need a third opinion on this one :smile:

Reply:
Conditions should be fixed before the reconciliation happens, see KRShaped thing in pkg.

Reply:
We can change that but we lose some useful features.

Reply:
We can emit an event in the CO context, but the general recommendation in the no-op case is to not emit events: https://github.com/knative/pkg/issues/1520, and today we don't really differentiate if we actually create or not the topic in Kafka since we suppress the `TopicAlreadyExistsError`.

Reply:
> On the contrary, to me this really sounds like a status thing: TopicAlreadyPresent vs NewTopicCreated

If this is the case the distinction should be in the Kafka channel as well, but there isn't such distinction: https://github.com/knative/eventing-contrib/blob/9547d5559586095acfbec6ffa39e8e52d6aee004/kafka/channel/pkg/apis/messaging/v1beta1/kafka_channel_lifecycle.go#L27

Reply:
Let me know what do you think, I just wanted to point out the implication of such a change.

Reply:
@pierDipi in the case of the channel, the distinction doesn't exist because the controller always owns the topic

Reply:
right, the topic of the channel is a different beast: it's more for backing the messages flying in

Source (and Sink), is totally different.  In most cases the topic will be created by an "external" entiy - most production clusters even disallow the creation of a topic if it does not exist.

Reply:
Agree with @matzew . I think you can log who owns the topic in the condition of topic ready and that should be enough for the user

Reply:
I misunderstood your comment, you meant the reason and not a different condition, I was confused by NewTopicCreated vs ...., my bad :), and the channel example was a really bad example.